<h1> To remove duplicate images from a given dataset </h1>
<ol>
    <li> Average Hashing    (aHash) </li>
    <li> Peceptual Hashing (pHash) </li>
    <li> Difference Hashing (dHash) </li>
</ol>

In [20]:
import cv2
import os
import numpy as np
import matplotlib.pyplot as plt
from scipy.fftpack import dct
import csv

# using aHash

1. Import Images
2. Reduce there sizes to 8x8
3. convert image to grayscale
4. Average the colors
5. Based on whether an individual pixel more than average value , we assign it a 1/0 value

In [15]:
# Importing images

img = []   # Names of the images
names = [] # Names of image files
ctr =0
for dog in os.listdir('aug_dogs'):
    names.append(dog)
    ctr+=1
    filePath = os.path.join('aug_dogs/',dog)
    dogImg = cv2.imread(filePath)
    #Converting image to gray scale
    dogImgGrayScale = cv2.cvtColor(dogImg,cv2.COLOR_BGR2GRAY)
    #Resizing image
    dogImgResized = cv2.resize(dogImgGrayScale,(8,8))
    img.append(dogImgResized)

In [16]:
#Computing hash values for each img
aHash  =[]
for dog in img:
    mean_color = np.mean(dog)
    new_mat = [x>mean_color for x in dog]
    new_mat = np.array(new_mat,dtype = 'int').reshape((-1,))
    binVal = ''.join(str(x) for x in new_mat)
    hashVal = hex(int(binVal, 2)) # for reference
    aHash.append(binVal)

In [17]:
# Making a dictionary of values having same bits
duplicated = {}
for i in range(len(names)):
    curr_name = names[i]
    curr_hash = aHash[i]
    try:
        duplicated[curr_hash].append(curr_name)
    except:
        duplicated[curr_hash] = []
        duplicated[curr_hash].append(curr_name)

In [34]:
# Analyzing the results
duplicated_names = 0
duplicated_items = 0
total_images =0 
for keys in duplicated.keys():
    total_images+=1
    if(len(duplicated[keys]) >= 2):
        duplicated_names+=1
        duplicated_items += len(duplicated[keys])
        print(duplicated[keys])
# # Saving the result
with open('aHash.csv', 'w') as f:
    for key in duplicated.keys():
        for elem in duplicated[key]:
            f.write("%s,"%(elem))
        f.write("\n")

['c_dog.10.jpg', 'dog.10.jpg', 'n@dog.10.jpg']
['c_dog.100.jpg', 'dog.100.jpg']
['c_dog.1000.jpg', 'dog.1000.jpg', 'n@dog.1000.jpg']
['c_dog.1001.jpg', 'dog.1001.jpg']
['c_dog.1002.jpg', 'dog.1002.jpg', 'n@dog.1002.jpg']
['c_dog.1003.jpg', 'dog.1003.jpg']
['c_dog.1004.jpg', 'dog.1004.jpg', 'n@dog.1004.jpg']
['c_dog.1005.jpg', 'dog.1005.jpg', 'n@dog.1005.jpg']
['c_dog.1006.jpg', 'dog.1006.jpg']
['c_dog.1007.jpg', 'dog.1007.jpg']
['c_dog.1008.jpg', 'dog.1008.jpg', 'n@dog.1008.jpg']
['c_dog.1009.jpg', 'dog.1009.jpg', 'n@dog.1009.jpg']
['c_dog.101.jpg', 'dog.101.jpg']
['c_dog.1010.jpg', 'dog.1010.jpg']
['c_dog.1011.jpg', 'dog.1011.jpg']
['c_dog.1012.jpg', 'dog.1012.jpg']
['c_dog.1013.jpg', 'dog.1013.jpg']
['c_dog.1014.jpg', 'dog.1014.jpg']
['c_dog.1015.jpg', 'dog.1015.jpg']
['c_dog.1016.jpg', 'dog.1016.jpg', 'n@dog.1016.jpg']
['c_dog.1017.jpg', 'dog.1017.jpg']
['c_dog.1018.jpg', 'dog.1018.jpg']
['c_dog.1019.jpg', 'dog.1019.jpg']
['c_dog.102.jpg', 'dog.102.jpg', 'n@dog.102.jpg']
['c_dog.102

In [22]:
print(duplicated_items,duplicated_names,total_images)

2383 1001 6432


<p> Our analysis using aHash reveals that there are a total of 1001 uniques images(out of 6432 unique images) which have been duplicated a grand total of 2383 </p>

In [80]:
# Visualizing duplicates
itr = 0
for keys in duplicated.keys():
    if(itr == 10 ):
        break
    if(len(duplicated[keys]) >= 2):
        itr+=1;
        print("Encountered Duplicates")
        for values in duplicated[keys]:
            filePath = os.path.join('aug_dogs/',values)
            curr_img = cv2.imread(filePath)
            cv2.imshow(str(itr),curr_img)
            cv2.waitKey(0)
            cv2.destroyAllWindows()

Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates


# using pHash

1. Import Images
2. Reduce there sizes to 32x32
3. convert image to grayscale
4. Convert image to its dct eqv
5. Crop the upper 8x8 half
6. Take average of pixel values
7. Iterate through all the pixels and if they are pixel values > average then make them 1 else 0

In [40]:
# Importing images

img = []   # Names of the images
names = [] # Names of image files
ctr =0
for dog in os.listdir('aug_dogs'):
    names.append(dog)
    ctr+=1
    filePath = os.path.join('aug_dogs/',dog)
    dogImg = cv2.imread(filePath)
    #Converting image to gray scale
    dogImgGrayScale = cv2.cvtColor(dogImg,cv2.COLOR_BGR2GRAY)
    #Resizing image
    dogImgResized = cv2.resize(dogImgGrayScale,(32,32))
    img.append(dogImgResized)

In [65]:
#Computing hash values for each img
pHash  =[]
for dog in img:
    
    dctImage = dct(dog,type = 2)[:8,:8]
    upperHalfDct = (dctImage.sum() - dctImage[0][0])/63.0
    dctImage = dctImage.reshape((-1,))
    dctImage = [x>upperHalfDct for x in dctImage]
    dctImage = np.array(dctImage,dtype = 'int')
    dctImage = dctImage.reshape((-1,))
    binVal = ''.join(str(x) for x in dctImage)
    hashVal = hex(int(binVal, 2)) # for reference
    pHash.append(binVal)

In [66]:
# Making a dictionary of values having same bits
duplicated = {}
for i in range(len(names)):
    curr_name = names[i]
    curr_hash = pHash[i]
    try:
        duplicated[curr_hash].append(curr_name)
    except:
        duplicated[curr_hash] = []
        duplicated[curr_hash].append(curr_name)

In [67]:
# Analyzing the results
duplicated_names = 0
duplicated_items = 0
total_images =0 
for keys in duplicated.keys():
    total_images+=1
    if(len(duplicated[keys]) >= 2):
        duplicated_names+=1
        duplicated_items += len(duplicated[keys])
        print(duplicated[keys])
# # Saving the result
with open('pHash.csv', 'w') as f:
    for key in duplicated.keys():
        for elem in duplicated[key]:
            f.write("%s,"%(elem))
        f.write("\n")

['c_dog.10.jpg', 'dog.10.jpg']
['c_dog.100.jpg', 'dog.100.jpg']
['c_dog.1000.jpg', 'c_dog.3137.jpg', 'dog.1000.jpg', 'f0@dog.1816.jpg', 'f1@dog.1408.jpg', 'n@dog.1000.jpg']
['c_dog.1001.jpg', 'c_dog.1002.jpg', 'c_dog.1004.jpg', 'c_dog.1014.jpg', 'c_dog.1018.jpg', 'c_dog.1023.jpg', 'c_dog.1046.jpg', 'c_dog.1048.jpg', 'c_dog.105.jpg', 'c_dog.1057.jpg', 'c_dog.1058.jpg', 'c_dog.1064.jpg', 'c_dog.1065.jpg', 'c_dog.107.jpg', 'c_dog.1076.jpg', 'c_dog.1079.jpg', 'c_dog.1083.jpg', 'c_dog.1090.jpg', 'c_dog.1091.jpg', 'c_dog.1095.jpg', 'c_dog.11.jpg', 'c_dog.1100.jpg', 'c_dog.1112.jpg', 'c_dog.1113.jpg', 'c_dog.1114.jpg', 'c_dog.1117.jpg', 'c_dog.113.jpg', 'c_dog.1132.jpg', 'c_dog.1140.jpg', 'c_dog.1146.jpg', 'c_dog.1156.jpg', 'c_dog.1162.jpg', 'c_dog.1169.jpg', 'c_dog.1173.jpg', 'c_dog.1195.jpg', 'c_dog.1206.jpg', 'c_dog.1208.jpg', 'c_dog.1210.jpg', 'c_dog.1214.jpg', 'c_dog.1216.jpg', 'c_dog.1221.jpg', 'c_dog.1226.jpg', 'c_dog.1242.jpg', 'c_dog.1243.jpg', 'c_dog.1257.jpg', 'c_dog.1270.jpg', 'c_

In [69]:
x =['c_dog.1000.jpg', 'c_dog.3137.jpg', 'dog.1000.jpg', 'f0@dog.1816.jpg', 'f1@dog.1408.jpg', 'n@dog.1000.jpg']
for item in x:
    curr_img = cv2.imread('aug_dogs/' + item)
    cv2.imshow('img',curr_img)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

In [68]:
print(total_images,duplicated_names,duplicated_items)

3746 853 4921


Acc to pHash , there are a total of 3746 unique images with 853 duplicated images over 4921 times
<strong> 
    This Does not look correct as I have visualized the results and they dont look correct( However there is a change that i could have analysed the misclassifications )
</strong>
This Could be corrected if we use similiarity between hamming codes , but that would help us in finding whether two images are similar or not , it will not help us in grouping all the images which are duplicates

or it could be that pHash tries to find generalization in images

# using dHash

1. Import Images
2. Reduce there sizes to 9x8
3. convert image to grayscale
4. Averaging the color
5. from index 1 .. 8 compare each cell in each row , with its previous cell in the same row , assign bit values on the relative magnitude of these 2 values

In [35]:
# Importing images

img = []   # Names of the images
names = [] # Names of image files
ctr =0
for dog in os.listdir('aug_dogs'):
    names.append(dog)
    ctr+=1
    filePath = os.path.join('aug_dogs/',dog)
    dogImg = cv2.imread(filePath)
    #Converting image to gray scale
    dogImgGrayScale = cv2.cvtColor(dogImg,cv2.COLOR_BGR2GRAY)
    #Resizing image
    dogImgResized = cv2.resize(dogImgGrayScale,(9,8))
    img.append(dogImgResized)

In [36]:
#Computing hash values for each img
dHash  =[]
for dog in img:
    
    x1 = dog[:,0:8]
    x2 = dog[:,1:9]

    gradient = np.array((x1>x2),dtype = 'int')    
    gradient = gradient.reshape((-1,))
    binVal = ''.join(str(x) for x in gradient)
    hashVal = hex(int(binVal, 2)) # for reference
    dHash.append(binVal)
    print(hashVal)



0xb5da545ad94d3769
0x6ad8ddcdaa4a76f6
0xc6abd792ddaca9c6
0x158353e56dcdcd34
0xcd9f862b53cb858d
0xaa54d4687cce5454
0xe7cdeba7d7d79717
0x2d0a26a64ed79394
0xe05574db262b29a6
0x8c175b472f6b1f0d
0x849358b2b2b4ba32
0x5f7b6767d5d0e9ea
0xbc1e0e5a578f8fc7
0xdacdceeae0b2c5e6
0xc59ab7b36aa642b
0x57906164e42e6c66
0x67230a3c4e64168e
0xd5e95d5375753aca
0x4714d68cda5a5af5
0x493665252cd562ad
0x7cb9dbc98d081391
0x1533865e174ed56e
0x5362f2a8a6325a6b
0x753a1c36d3edb2ae
0xdba75354a12541a5
0x5c7caadcb0b2d576
0x8a306623732a6830
0x4243ae2a542a36eb
0x11255470a415954c
0x68ece2d3d42a8f97
0xbf9cac4cc94cb51b
0xdab26459d1d12284
0xe7b499a54b6d4a2f
0x8babc3a9aa6b2f2b
0x6455ba0a590c0faa
0x3a4f474e2e3a57b6
0x2ec7d6c59d1a853f
0x7318598a4aa2868e
0x6676a75637275517
0x13ad93072d612b73
0xed95a99575512244
0x93ccd472a2a6a599
0xb5f5d47050e56068
0x5be57168a8e9ad6a
0x2705054b372f3ebd
0xac54dc6ce958d1c5
0x3127d6469f1e484c
0xcb53cf2713290183
0x268ce9ceeedd9555
0xf5b656aa1b5391d
0x6e4d83ceb6b4252c
0x6b1f5f1a994bab57
0xcacd94c78793

0x3dafc92dcbabcb1f
0xe0a8e422307a4a6e
0x9836d552333569ac
0x2ad8d4de1e948a64
0x5629ef468f1d2db1
0x67b3a3a390154505
0x2aa6b5b7b9b7139a
0xa3e36561e58dac8a
0x13da52cc4c658d05
0xe66cdab2a4a052b6
0x3a7df872ca5c4cd0
0x2595bada923a6bbc
0x3774a9aca6263bd2
0x49716bd6b9d24c38
0x5c3ca4e8aa656e45
0x90d6725151b25339
0x7a3068cadad6f3a1
0xe0e46ac4d3c90738
0x469eb4f4bae2da99
0x122b33f45c95275f
0x89cce47274214d11
0x49c9ebd3a3554276
0x2364f4ede63241a5
0xa275ae2e5293272a
0xadada9fc7816b657
0x834cc428b088d154
0x8bb3336773e9eafb
0x94c7b4c4ce0e271f
0x2a9e0ddb9f86961d
0xf1dcd65c56fab7cd
0x352562928196249a
0x1c195011b02048e0
0x450465a6b3b2703f
0x31733b33232b6b3b
0x7a2a2f274e5a2936
0x2adcc82ca90a6ba8
0xd726aee8ab942657
0x9cab4b773333135a
0xa9e2f8f07570700a
0x554279ae37b25a0d
0x934da8aead4ce8ee
0x2eba3343d6aeabad
0xb669c74f0b23ba9
0x8e03575c1f2e74e7
0x53b1050901970e16
0x9261b0b20a618551
0x4c615399d55ab56a
0x56a42e3c3cb42c14
0x69cdb84868645550
0xbcd7eed7c1759d66
0x99b4a6f267523d2e
0x2e5929b2995a9a5a
0xa9abe3b5d85

0x63e35b4b919d8519
0xa22d242c3dddfb57
0x24924b6662b36c22
0x9e5f4b539955d599
0xca5b8d57559a8be1
0xef17564d46cf2532
0xd2f38a75c62bca12
0x3b3a2bb79b699a89
0x4e2f994d2e6a69aa
0xaca7b2b8b2a8a0d5
0x6d710e3c3b992c1d
0x57337996a3a4f9b2
0x68333609095859e9
0xf0988cfbed64f675
0xb2f1c1e1f571323
0xb4ce8ba48b9362d4
0x59612a45abacaa2e
0x96d5d78e9d9f55e4
0x9125afb43854574a
0xbf1e9eb88ac14786
0x8fb0746838b45872
0xc8c655898be653ab
0xa0153c72c7d47497
0x5981c8de6abbce0d
0x3b119d93556626f4
0x535aad8d478fa121
0xc223a9dc94bb3251
0x3dfad486c3051b17
0x57ad45bad2322ab6
0x9a7a5c89be3d7633
0xd45c34961735342c
0x97b079c3a2a55371
0x225978648de4db27
0x493ab0f8f1dcecf4
0xd496350e4e1d1d1c
0x9e4de2656eaa231b
0x75c8aaa42a24322a
0x2525a6162fdb0b17
0x3b98686871942350
0x593aeca4d3a485ba
0xe092c646c8aaa2b2
0x2c12241a19aab797
0xa586172696224971
0x3d67456d4755d6a6
0xdac9ab97cbcb930f
0x351599a97361e3e6
0x9aaf578faa09956d
0xc94c53191d199544
0x1e191e141435152f
0x33f7b5e544adaa68
0xf6f2eec692e7dcaa
0xdad6cb121e56955d
0x1124252a8a5

0xeaa76a8a12a21d1c
0x536dc964a4d34b45
0xc23ce4b19553513
0x59b3144701052e4c
0xa3255466542637ab
0x58ead175c3c59a9d
0x494b4f67d9f496ae
0x862265270d948d05
0x624a4a0696325297
0x2fba9a8d4d2d77b5
0x420c8142d2c7661d
0x8c12705633217477
0xad2a642d3d9869ce
0xa4812840684850a4
0xab8aaaf2fad0520b
0x3938252d76465e67
0xba4b9bcb43442015
0xdbb62d11b932c9ab
0x84052d3466555466
0xac254e15d5a62e38
0x817954d9d0c5bb3
0xb430ae519c2bacb7
0x23a32b122669266e
0x8d0b624d8ba969ba
0xca1ab2e2b2eafa54
0x49718fc32366cb47
0x153361963ada60b2
0xe933936d6fe56568
0xf0e6ce2008c99051
0x2f0bc7860715173b
0xd28c2eda2e36b9d4
0x6579a35d2acaaa8b
0x9654148e460655d8
0x765b0ad2e2d515ad
0x28786e2ae7c1d8e
0xe7251e9e3d2e5b1
0xcc9c556e2e98342a
0xe857c3a11cd4d016
0x657aec84a9228c4f
0x2357463655b99bd0
0x35a54a4e1d0e7a7b
0xbc3b6ac4d622b335
0x40a0d49e2c5f2717
0x154a55a2b4b3ab92
0xa6a1c56e82019133
0xd4c5d3961753d3cb
0x16d2612cba5a3561
0x3b65e1d94ed8241b
0x6da3f2e070c4c8d0
0xd496538f8d4747c7
0x44db85989aa3b27
0x51ecaadaabdbb3ab
0x5b5b9a970b242f1

0x8bcac7c76e6f6f5f
0x7c5d302955a16289
0x419492998e8b890a
0x80c2678d26371775
0xb61b272f4b72c707
0xccc1c99dded72b
0xab171bf3715496fb
0xb785cba54ab9ab15
0x8e11aa0206a635ed
0x48f838f2d0411009
0x93288664b8ba90d
0x7729cac8ea3aaa70
0x66a4b4ab73b10923
0xfa5b65a01a8540c
0xc8ece950bbc23469
0xda8086d2f07270e
0xecd2d69eb6bab6ce
0x5a593165b452684b
0xf9bb9f141c9cafad
0xf71cbeb2ea6a3a3a
0x856851515d38d11
0x99da8e969acdb2d6
0xc2546692152a6121
0xe6cc963e5e6eaa16
0x6c342c554440ac4a
0x10b08fe4ccc84cb9
0xd8a554a0bb9b9b1
0xf0b67c684150eaee
0xe9cba3a98d483020
0x59d04a4961d54d65
0xa2c5d160d4656e73
0x8cd6d9cb51cbc9ca
0x8a56aabb32aa3b8a
0x11d4a268b49a6a8e
0x9243169e971a25e9
0xbc5a8a62b45834b5
0xb0d97926232b1c1f
0xa5b6ae539b8fc6b1
0x6a5c569a4cadd934
0xac8a72cccd6d656e
0x36345bab0c5995a7
0xe2d2f8fcacbd620
0x854645cb4963aba5
0x9c9cb68e8ebe9fa
0x384df5d49ce3650d
0x85a62b195b4b32b2
0x3591ede2a594d455
0x818c4c3178760f1b
0xd97260646029e8f8
0x98e9ecad969463a8
0x31ccd44a0ec82d44
0x562dce72d43a5a8a
0x958ec517a7227396
0x

0xfdfc27c40bd64de4
0x5d11152e5a5a496c
0x13af532d234e2cbb
0xc2c0eadc5cbdd975
0x24545c5ab8787910
0xab558adc27b63119
0x10148347438181b
0x4d60b1f65e5f6f97
0x1e5665431b0f1f0f
0xd9b405c5196629e5
0x3291e03af0f1b544
0x430861e4e5278a26
0x45c4b6b4c879a458
0xe175c70fcbbbab6f
0x9008989ad8caa0b0
0x1d1f2b6e4d894dbc
0xc5a189487acb064d
0x9234ea76b5d651b6
0x6d8c87850726aad9
0xca4466220000
0x6a734936268b0980
0x22a232d54dd99333
0x1e488823851931a1
0x1e1a0c386aa8a90b
0xbabcdbe6b173dd1
0x36aa3d5d96ce1ee5
0x4a5f2c0e92d35000
0xebcad21a5255abef
0xb5c563aaabd7abb1
0x28b0f0d8a09830a8
0xa6a5b4cb9a48fbd4
0x2b9a970191952d93
0x4eb65d939196951e
0xdb444f4fbebd3a6d
0x3032acc490d86e03
0x92d433717b9d1e8f
0x7ee83268389258e0
0xaa7795a39856d9d5
0x1725b5e6a870e331
0x4f86068b697000
0x842d4c2e4a0a0b63
0x492e532cf4f83950
0x54d652da9ab16585
0xd76faac7ed5b2a55
0x49649ab455ceb83e
0x13696153bb4953e9
0x121ac7da95f5d935
0xc645558cd416b25a
0xf69c87b55bcbb9e9
0x44522e66b75efefe
0x8db237bacd4b131a
0xdcdfeceeb2fbb46d
0x9bf1b5e9fa594c4c
0

In [37]:
# Making a dictionary of values having same bits
duplicated = {}
for i in range(len(names)):
    curr_name = names[i]
    curr_hash = dHash[i]
    try:
        duplicated[curr_hash].append(curr_name)
    except:
        duplicated[curr_hash] = []
        duplicated[curr_hash].append(curr_name)

In [38]:
# Analyzing the results
duplicated_names = 0
duplicated_items = 0
total_images =0 
for keys in duplicated.keys():
    total_images+=1
    if(len(duplicated[keys]) >= 2):
        duplicated_names+=1
        duplicated_items += len(duplicated[keys])
        print(duplicated[keys])
        
# # Saving the result
with open('dHash.csv', 'w') as f:
    for key in duplicated.keys():
        for elem in duplicated[key]:
            f.write("%s,"%(elem))
        f.write("\n")

['c_dog.10.jpg', 'dog.10.jpg']
['c_dog.100.jpg', 'dog.100.jpg', 'n@dog.100.jpg']
['c_dog.1000.jpg', 'dog.1000.jpg']
['c_dog.1001.jpg', 'dog.1001.jpg']
['c_dog.1002.jpg', 'dog.1002.jpg']
['c_dog.1003.jpg', 'dog.1003.jpg']
['c_dog.1004.jpg', 'dog.1004.jpg']
['c_dog.1005.jpg', 'dog.1005.jpg']
['c_dog.1006.jpg', 'dog.1006.jpg']
['c_dog.1007.jpg', 'dog.1007.jpg']
['c_dog.1008.jpg', 'dog.1008.jpg']
['c_dog.1009.jpg', 'dog.1009.jpg']
['c_dog.101.jpg', 'dog.101.jpg']
['c_dog.1010.jpg', 'dog.1010.jpg']
['c_dog.1011.jpg', 'dog.1011.jpg']
['c_dog.1012.jpg', 'dog.1012.jpg', 'n@dog.1012.jpg']
['c_dog.1013.jpg', 'dog.1013.jpg']
['c_dog.1014.jpg', 'dog.1014.jpg']
['c_dog.1015.jpg', 'dog.1015.jpg']
['c_dog.1016.jpg', 'dog.1016.jpg']
['c_dog.1017.jpg', 'dog.1017.jpg']
['c_dog.1018.jpg', 'dog.1018.jpg']
['c_dog.1019.jpg', 'dog.1019.jpg']
['c_dog.102.jpg', 'dog.102.jpg', 'n@dog.102.jpg']
['c_dog.1020.jpg', 'dog.1020.jpg']
['c_dog.1021.jpg', 'dog.1021.jpg']
['c_dog.1022.jpg', 'dog.1022.jpg']
['c_dog.1023.

In [39]:
print(total_images)
print(duplicated_names)
print(duplicated_items)

6700
1001
2115


A total of 1001 images out of 6700 have been duplicated a total of 2115 times

<h3> Conclusions </h3>
<hr>
1.aHash -  Our analysis using aHash reveals that there are a total of 1001 uniques images(out of 6432 unique images) which have been duplicated a grand total of 2383 

2.pHash - Acc to pHash , there are a total of 3746 unique images with 853 duplicated images over 4921 times This Does not look correct as I have visualized the results and they dont look correct( However there is a change that i could have analysed the misclassifications ) This Could be corrected if we use similiarity between hamming codes , but that would help us in finding whether two images are similar or not , it will not help us in grouping all the images which are duplicates

3.dHash - A total of 1001 images out of 6700 have been duplicated a total of 2115 times